aspnetcore/mvc/models/model-binding.md

```diff
@@ -857,8 +857,8 @@ To make the ASP.NET Core route value provider and query string value provider un
 * Replace the [culture value](https://github.com/aspnet/AspNetCore/blob/e625fe29b049c60242e8048b4ea743cca65aa7b5/src/Mvc/Mvc.Core/src/ModelBinding/QueryStringValueProviderFactory.cs#L30) passed to the value provider constructor with [CultureInfo.CurrentCulture](xref:System.Globalization.CultureInfo.CurrentCulture)
 * Replace the default value provider factory in MVC options with your new one:
 
-[!code-csharp[](model-binding/samples_snapshot/2.x/Startup.cs?name=snippet)]
-[!code-csharp[](model-binding/samples_snapshot/2.x/Startup.cs?name=snippet1)]
+[!code-csharp[](model-binding/samples_snapshot/Startup.cs?name=snippet)]
+[!code-csharp[](model-binding/samples_snapshot/Startup.cs?name=snippet1)]
```

Comment:
This code is for 3.0, but it's shown in both versions. I removed `2.x` from the folder structure to emphasise that.

Should we?:

1. Create a 2.2 version of this code and reference that accordingly.
2. Remove this section from the 2.2 version.
3. Do nothing.

Reply:
I think it depends on the differences for what's being shown. If churn occurred (as is usually the case for 2.x and 3.x in *Startup.cs*), a good approach is to create a *3.x* folder in the snapshots folder to hold the 3.x *Startup.cs* in parallel to the 2.x version of the file. That way, the future versions (5.x, 6.x ...) can follow suit, and the topic's links are easy to manage over time (i.e., **_saves MS 💰_**). Surely tho, it's overkill when there haven't been any changes between 2.x and 3.x and there won't likely be changes in the future. I almost always just dump those right into the root of the *samples_snapshots* folder.

Reply:
I just took a look at the diff. I don't think this is the best way. Version the sample code from each sample, even if there aren't changes. Also note that the PU has stated that they really dislike the *full topic versioning* approach. It makes topic updates *very hard for them to review*. I've also noted that it's a pain point to mod every section of the ones that duplicate the whole topic up to three times.

```
::: moniker range=">= aspnetcore-3.0"

3.x sample link here

::: moniker-end

::: moniker range="< aspnetcore-3.0"

2.x sample link here

::: moniker-end
```

Reply:
... but I'll add that it has been a *necessary evil* at times. Some of the topics *wildly flipped* in and out of versioning blocks ... those had to use *full topic versioning*. For example, Kestrel. *ei ei ei* ... it was a 🚂💥 before we changed it over to *full topic versioning*.

Reply:
Thanks @guardrex. That's very valuable feedback. I try to make decisions based on things I see elsewhere in existing samples, recent PRs, etc, but I'll admit that can be problematic. It's not always clear what the current thinking is and naturally different samples do differing things.

If you have the time, can you turn this into a task-list for me? You can dial the nit level up to _crazy_ too. I'm not sure what to do with this now.

Reply:
I’m at the gym now 💪. When I get back to the house, I’ll see if I can find the original issue where the sample layout decisions were made. You can look at mine btw on topics that I wrote. They’re highly consistent with that guidance. Ping u back in a bit.

Reply:
Btw just for fun on the subject of consistency and guidelines ... it’s been a long-standing concern of mine. See 👉 https://github.com/aspnet/AspNetCore.Docs/issues/2617

Reply:
I typically make a 3.x copy of the code. If there are no changes, just use the 2.x code and don't change the path.

Reply:
I *think* I've tracked down the very first discussion about sample layouts ... it seems to start with my remark here and follows with a discussion ...

https://github.com/aspnet/AspNetCore.Docs/pull/3896#issuecomment-322605262

I noticed that the PR's that I put in just after that ... several sample updates in late August of 2017 ... were the first to use it.

@serpent5, of course, follow Rick's advice. I'd prefer to stick with what I've been doing. I think it's the fastest 🕐, least confusing 😵, and easiest ❤️ way to go ... i.e., I think it's the lowest cost **_$$💰$$_** approach in the long run.

I'll try to use the *whole topic versioning* only when really desperately needed given my own hardships with it and the complaints from engineers about it.

I'll make a quick pass on my old 🐮 **_BEEF!_** 🐄 issue just in case I need to call it out again. It will only take a few minutes to fix up.

Reply:
I'll mention one other change tho since then. Back in those days, I think we felt that samples should named (the project folder should be named) with something specific to the sample (e.g., *ConfigurationSample* or *ChangeTokensSample*). I liked that (still do) because when I have multiple samples open, it's easier to find them in tabs and tooltips in IDEs.

Later tho ... after a bunch of samples were named that way, the guidance shifted to naming all sample apps *SampleApp*. IIRC, the reasoning had to do with cut-'n-paste, but I don't think ever felt too good about that ... devs are going to have to change namespaces no matter what.

It's only really a problem in those rare cases where you have more than one sample at each version ... for example a console app sample and an ASP.NET Core sample in each of the 2.x and 3.x folders. Otherwise, they would be called something like *SampleApp1* and *SampleApp2*, which is kind'a funky. Anyway ... that's *very rare*.

Just calling that out for you @serpent5: The guidance is to call every sample app (project folder) *SampleApp*. You don't have to change them; but if you were creating a **_new_** topic+sample(s) from scratch, that's the best way to go.

Reply:
I've never heard the SampApp advice. I frequently use WebApp1 or something related to the sample - WebConfiguration

Reply:
You told me to do that. :smile: lol

Reply:
I can look it up if you want ... but it had to do with namespaces and cut-'n-paste IIRC. I think you said it would be easier for devs to consume our code bits in their own apps. It was a long time ago that it was discussed. I've been making all **_new_** samples use *SampleApp*.